### PR TITLE
NV-ModelOPT INT4 quantization

### PR DESCRIPTION
**Pull request by riyadshairi979**
## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

## Tests
- Unit test: `pytest test/unit_test/passes/onnx/test_nvmo_quantization.py`
- Example: `python -m olive.workflows.run --config bert_nvmo_ptq.json`

### CODE REVIEW

File: olive/passes/onnx/nvmo_quantization.py
Comment (jambayk) [inline olive/passes/onnx/nvmo_quantization.py:75]: this default value doesn't match the searchable values. Also, the parameter might be better named as "precision"?

File: olive/passes/onnx/nvmo_quantization.py
Comment (jambayk) [inline olive/passes/onnx/nvmo_quantization.py:54]: since this param is not used in the pass, it can be removed

File: examples/bert/bert_nvmo_ptq.json
Comment (jambayk) [inline examples/bert/bert_nvmo_ptq.json:34]: "metric" is also not used as well.

File: test/unit_test/passes/onnx/test_nvmo_quantization.py
Comment (jambayk) [inline test/unit_test/passes/onnx/test_nvmo_quantization.py:1]: Please also add the required package to https://github.com/microsoft/Olive/blob/main/test/requirements-test.txt.

File: olive/passes/onnx/nvmo_quantization.py
Comment (jambayk) [reply olive/passes/onnx/nvmo_quantization.py:75]: Oh it looks like only int4 is support right now. If so, this field can be removed as well.

Comment (jambayk) [reply olive/passes/onnx/nvmo_quantization.py:75]: "algoritm is not used as well

File: test/unit_test/passes/onnx/test_nvmo_quantization.py
Comment (jambayk) [reply test/unit_test/passes/onnx/test_nvmo_quantization.py:1]: looks like the `nvidia-modelopt` is only available on windows, so I think we have to disable the unit test entirely for now. We currently don't have separate test requirements based on os.

**Comment by riyadshairi979** [comment]
> @riyadshairi979 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree [company="Nvidia"]

**Comment by riyadshairi979** [comment]
@microsoft-github-policy-service agree company="Nvidia"

File: test/unit_test/passes/onnx/test_nvmo_quantization.py
Comment (jambayk) [reply test/unit_test/passes/onnx/test_nvmo_quantization.py:1]: @riyadshairi979 can you remove the package from the test requirements because it will fail to find the package on Linux. 
Instead please add a `@pytest.mark.skip(reason="nvidia-modelopt not available on Linux. Re-enable after introducing OS specific requirements")` instead of the current `skipif`

Comment (riyadshairi979) [reply test/unit_test/passes/onnx/test_nvmo_quantization.py:1]: Actually [modelopt](https://pypi.org/project/nvidia-modelopt/) is going to be available for both windows and linux on next Monday/Tuesday. I will remove the `skipif` in next commit.

Comment (jambayk) [reply test/unit_test/passes/onnx/test_nvmo_quantization.py:1]: Oh that's great! We won't be able to run the CI or merge it until then but I can review the logic.

File: olive/passes/onnx/nvmo_quantization.py
Comment (jambayk) [inline olive/passes/onnx/nvmo_quantization.py:108]: will other precisions be supported in the future? If not, then we could just remove the "precision" config parameter.

Can you add this check as part of `validate_search_point` like in https://github.com/microsoft/Olive/blob/3120697d4e9d2dee5af0877d9cfc082f6a422dfb/olive/passes/onnx/optimum_conversion.py#L53

Comment (jambayk) [reply olive/passes/onnx/nvmo_quantization.py:108]: It is also possible to use a `str,Enum` as type for the param which comes with automatic validation like in https://github.com/microsoft/Olive/blob/3120697d4e9d2dee5af0877d9cfc082f6a422dfb/olive/passes/onnx/model_builder.py#L71

File: test/unit_test/passes/onnx/test_nvmo_quantization.py
Comment (jambayk) [reply test/unit_test/passes/onnx/test_nvmo_quantization.py:1]: oh that's great! I will wait until the release to trigger the PR validation pipelines then.

File: olive/passes/onnx/nvmo_quantization.py
Comment (riyadshairi979) [reply olive/passes/onnx/nvmo_quantization.py:108]: Yes, other precisions will be supported as well. Looking into `validate_search_point `.

File: test/unit_test/passes/onnx/test_nvmo_quantization.py
Comment (riyadshairi979) [reply test/unit_test/passes/onnx/test_nvmo_quantization.py:32]: The `quantize_int4` API uses `CalibrationDataReader` but the other `quantize` API of modelopt uses iterable. We will make the change to make it congruent everywhere and compatible with Olive.

File: olive/passes/onnx/nvmo_quantization.py
Comment (jambayk) [inline olive/passes/onnx/nvmo_quantization.py:26]: Just some minor comments about the description of the config parameters. Please remove the references to `quant_mode` and `static`. 

If a calibration data loader is always needed, please add `required=True` under `dataloader_func` and update the descriptions of the other related params to remove the "only used if ..." part.

Can you also add a docstring under pass class to give a description of what this pass does? Thanks!

File: test/requirements-test.txt
Comment (jambayk) [inline test/requirements-test.txt:19]: Looks like onnx-graphsurgeon is also required. From the windows ci failure:
image

Comment (riyadshairi979) [reply test/requirements-test.txt]: I have updated the requirement with correct package.
qq: we have a dependency on `onnxruntime~=1.16.3`, will that conflict with onnxruntime requirement for other packages/tools within olive?

Comment (jambayk) [reply test/requirements-test.txt]: Oh, that is an issue. Our CI tests for the most recent stable release of ort. We have another pipeline that tests for the previous major release of ort but that one will soon migrate to 1.17.x since 1.18 is on the way.

Comment (riyadshairi979) [reply test/requirements-test.txt]: How to see the status of the pipeline? The `int4` support that we added in this PR does not require strict `1.16.3`, it will work with latest ort too. Let see if pipeline passes.
We will update our ort dependency to latest version for other quantization formats like `fp8`, `int8` before supporting them in olive.

Comment (jambayk) [reply test/requirements-test.txt]: okay, let me trigger the ci pipeline. you can see the status by clicking the jobs that show up. it takes you to another page that has links to the azure pipeline jobs.

Comment (jambayk) [reply test/requirements-test.txt]: I see the the `[all]` extra pins the ort version so this will reinstall ort. Can we directly install onnx-graphsurgeon package so that the ort package is not reinstalled? 
image

Comment (riyadshairi979) [reply test/requirements-test.txt]: I think it is doable, I have updated the requirements.
Bdw, I do not see `onnxruntime` as dependency in requirements-tests.txt, is it coming through core lib requirements?

Comment (jambayk) [reply test/requirements-test.txt]: it is installed separately so that we can control the version being tested  https://github.com/microsoft/Olive/blob/63c0430fe3da033e086e1dfbe00d9bdc918ecca8/.azure_pipelines/job_templates/olive-test-template.yaml#L37

File: examples/bert/nv_user_script.py
Comment (jambayk) [inline examples/bert/nv_user_script.py:14]: have tested this example locally? the model_loader function is always called with `model_path` whether it is none or not

Comment (jambayk) [reply examples/bert/nv_user_script.py:14]: you can empty the cache directory when rerunning after making code changes. This can also be done by adding `"clean_cache": true` under `"engine"` section during dev. otherwise, it will just reload from cache if the config json did not change.

Comment (riyadshairi979) [reply examples/bert/nv_user_script.py:14]: I tested locally, it works.

Comment (riyadshairi979) [reply examples/bert/nv_user_script.py:14]: Sorry, there was a formatting error. Please start the pipeline one more time.

File: olive/olive_config.json
Comment (jambayk) [inline olive/olive_config.json:243]: will graphsurgeon always be required for modelopt passes? perhaps we can add it here too?

Comment (riyadshairi979) [reply olive/olive_config.json:243]: Yes, it will always be required. I will make the change after the current pipeline ends/crashes.

**Comment by riyadshairi979** [comment]
@jambayk By seeing the pipeline fails, just realized that the installed modelopt needs some fixes! We have to wait a day or two to get some of our int4 fixes from main branch to get into the release! Should we merge this PR by disabling the test or do you think we should wait?

**Comment by jambayk** [comment]
> @jambayk By seeing the pipeline fails, just realized that the installed modelopt needs some fixes! We have to wait a day or two to get some of our int4 fixes from main branch to get into the release! Should we merge this PR by disabling the test or do you think we should wait?

@devang-ml @EmmaNingMS what do you think?

**Comment by riyadshairi979** [comment]
Another question @jambayk: if modelopt adds dependency `onnxruntime-gpu` instead of `onnxruntime`, how it will differ in olive and what version it will install?

File: test/requirements-test.txt
Comment (jambayk) [inline test/requirements-test.txt:1]: @riyadshairi979 replying to https://github.com/microsoft/Olive/pull/1135#issuecomment-2098555504 here so that we can have a thread without having to quote reply. 
> if modelopt adds dependency onnxruntime-gpu instead of onnxruntime, how it will differ in olive and what version it will install?

The behavior depends on whether the dependency is a direct requirement or an extra. 
- The easiest case would be an extra since we can just not install using the extra and directly consume whatever ort is already installed in the environment
- If it's an extra but ort-gpu is still needed for the pass to run, we could skip the test on the cpu ci and only run it on the gpu ci
- It gets complicated if it's a direct dependency that gets installed when we install the modelopt package. This is because ort is shipped as many different packages (onnxruntime, onnxruntime-gpu, ort-nightly, ort-nightly-gpu) but they all install into the same `onnxruntime` namespace. this means that if `onnxruntime` is already installed, `onnxruntime-gpu` doesn't uninstall `onnxruntime` and just adds it's files into the `onnxruntime` package folder overwriting files if necessary. This could lead to unexpected behavior so we usually uninstall all ort packages before installing a new one. Since in the ci, we test for a specific version of ort, if the modelopt packages installs ort-gpu, we should add it directly to the `requirements-test-gpu.txt`. Even then we would have to separate the requirements between the stable and nightly pipelines (which installs and tests ort-nighlty)

tldr; the ideal scenario is that onnxruntime(-gpu) is an optional dependency/extra for modelopt. If the pass requires a specific flavor and version of ort, we skip the tests as needed.

Comment (riyadshairi979) [reply test/requirements-test.txt:1]: I see. So do you suggest we add only `onnxruntime` in our `modelopt` dependency and keep `onnxruntime-gpu` as optional?

File: olive/passes/onnx/nvmo_quantization.py
Comment (kevalmorabia97) [inline olive/passes/onnx/nvmo_quantization.py:111]:
```suggestion
                "Please install `olive-ai[nvmo]` or `nvidia-modelopt[onnx]` to use INT4 AWQ quantization!"
```

File: test/requirements-test.txt
Comment (riyadshairi979) [reply test/requirements-test.txt:1]: `pip install nvidia-modelopt` will NOT install those onnx requirements.

Comment (jambayk) [reply test/requirements-test.txt:1]: That will be fine then. The changes in this PR to `requirements-test.txt` is sufficient since the other `onnx` dependencies are already present.

File: test/requirements-test.txt
Comment (kevalmorabia97) [inline test/requirements-test.txt:20]:
```suggestion
nvidia-modelopt[onnx]~=0.11.0
onnx-graphsurgeon
```

File: test/requirements-test.txt
Comment (riyadshairi979) [reply test/requirements-test.txt:1]: cool. do you want me to move the `onnx-graphsurgeon` to `olive_config.json` [here](https://github.com/microsoft/Olive/pull/1135/files#diff-f359471e7bc949029f5a4508182e76085dcc0acbb8db0a8c171ca19051210689R243)?

Comment (jambayk) [reply test/requirements-test.txt:1]: yes. if graphsurgeon will always be required for this pass, adding it along with modelopt would be good.

Comment (jambayk) [reply test/requirements-test.txt:1]: btw, is onnxruntime-extensions also required for the pass?

Comment (riyadshairi979) [reply test/requirements-test.txt:1]: Not for INT4 support but for other quantization format we support plugins too, which requires `onnxruntime-extensions`.

**Comment by riyadshairi979** [comment]
Please have another pass @all

File: test/requirements-test.txt
Comment (jambayk) [inline test/requirements-test.txt:19]: can you change this to the previous `nvidia-modelopt~=0.11.0` and `onnx-graphsurgeon`. 
Like we discussed earlier, the `onnx` extra in the ci test would install the ort dependencies from this extra on top of what ort is already installed in the test environment.

File: olive/auto_optimizer/config_template/pass_capability.yaml
Comment (riyadshairi979) [inline olive/auto_optimizer/config_template/pass_capability.yaml:73]: @jambayk how this capability is used? ModelOpt work both with CPU and CUDA EP. But I didn't understand the `accelerator` field.

File: test/requirements-test.txt
Comment (jambayk) [reply test/requirements-test.txt:20]: This change cannot be made due to dependency clash between the ci's ort package requirements and the extras requirements. Please refer to the discussion above for more details https://github.com/microsoft/Olive/pull/1135#discussion_r1592736575

File: olive/auto_optimizer/config_template/pass_capability.yaml
Comment (jambayk) [reply olive/auto_optimizer/config_template/pass_capability.yaml:73]: Olive has a feature where it uses this yaml file to automatically create model optimization workflows by mixing and matching compatible passes. 
The accelerator field refers to the compute hardware device. Olive uses a combination of ep,accelerator to categorize workflows. so the eps and accelerators here are used to create a matrix of supported combos.

I think you can remove this pass from here. Since the pass does not support data configs yet, the auto optimizer would not be able to use this pass as it won't have the dataloader func to provide to it. 
@trajepl can confirm?

Comment (riyadshairi979) [reply olive/auto_optimizer/config_template/pass_capability.yaml:73]: I will remove for now. We can add them back later if needed.

File: olive/passes/onnx/nvmo_quantization.py
Comment (trajepl) [inline olive/passes/onnx/nvmo_quantization.py:34]: nitpick: btw, it is fine not to add data_config here. I am planning to implent a standalone data container to unify dataloader_func and data_config.

File: examples/bert/bert_nvmo_ptq.json
Comment (jambayk) [inline examples/bert/bert_nvmo_ptq.json:26]: Hi, I ran the example locally with the newly released linux package and it worked. 
image

Can you add `"opt_level":0` here? Otherwise, the transformers optimizer tries to use an inference session with the ep to do further session graph fusions. but that doesn't work if the ep is tensorrt as the session graph cannot be saved.

File: examples/bert/bert_nvmo_ptq.json
Comment (jambayk) [inline examples/bert/bert_nvmo_ptq.json:1]: Please also add a section for this workflow in the README.md. Thanks!

File: olive/passes/onnx/nvmo_quantization.py
Comment (riyadshairi979) [reply olive/passes/onnx/nvmo_quantization.py:108]: Good suggestion, thanks. I will make the change in next release.

File: olive/passes/onnx/nvmo_quantization.py
Comment (jambayk) [reply olive/passes/onnx/nvmo_quantization.py:34]: I think this was a follow up to another discussion about data configs. There is no change needed from your side at this point.

**Review by jambayk — DISMISSED** [review]
Thanks for the contribution! Will merge once the CI passes.

File: examples/bert/README.md
Comment (riyadshairi979) [inline examples/bert/README.md:104]: @jambayk should add something about deployment? like below
```Users can deploy the quantized ONNX model using TensorRT 10.x but that is not supported through Olive right now, stay tuned!`

Comment (jambayk) [reply examples/bert/README.md:104]: The comment about the deployment would probably be about ORT and not Olive? Since olive only produces the model and it gets deployed using ort, tensorrt or some other engine. I will let @EmmaNingMS comment on this.

Comment (EmmaNingMS) [reply examples/bert/README.md:104]: Vote for this version if ORT support is in the plan. "Deployment support for TensorRT-Model-Optimizer quantized models is coming soon in ORT, in the meantime try TensorRT 10.x"

What's the timeline to add the support in ORT? Will Nvidia contribute to it?

Comment (riyadshairi979) [reply examples/bert/README.md:104]: @EmmaNingMS I will ask team about the roadmap regarding these support. Thanks.

File: olive/passes/onnx/nvmo_quantization.py
Comment (jambayk) [inline olive/passes/onnx/nvmo_quantization.py:108]: Not a breaking issue since onnx opset 21 is not currently supported by ort. But I think the quantizer probably needs to update the model's opset to 21 if it is not already so? Or require the model's opset to be 21? (olive provides a pass for that too https://github.com/microsoft/Olive/blob/3e5ce3f6907eed5063d937af756e9a214a9342eb/olive/passes/onnx/conversion.py#L579)

Otherwise, the model graph is technically invalid for lower opset version since int4 in dq and q ops were only added in opset version 21. https://github.com/onnx/onnx/releases/tag/v1.16.0

Comment (riyadshairi979) [reply olive/passes/onnx/nvmo_quantization.py:108]: Our tool does not care much about opset version either. Anything else before we can merge this PR?

Comment (jambayk) [reply olive/passes/onnx/nvmo_quantization.py:108]: Sounds good to me. Everything looks good from my side. I will approve and merge once the CI passes. 

The only outstanding thing is the discussion about the deployment messaging but that can be updated in a follow up PR if necessary.

Comment (jambayk) [reply olive/passes/onnx/nvmo_quantization.py:108]: The failing/cancelled test is not related to this PR. I will rerun it once the other tests complete.

**Review by jambayk — APPROVED** [review]
Thanks!

**Pull request merged**
